bascule/key/cache.go

```diff
@@ -0,0 +1,210 @@
+		if pair, err := cache.delegate.ResolveKey(dummyKeyId); err == nil {
+			cache.store(pair)
+		} else {
+			errors = []error{err}
+		}
+	})
+
+	return
+}
+
+// multiCache uses an atomic map reference to store keys.
+// Once created, each internal map instance will never be written
+// to again, thus removing the need to lock for reads.  This approach
+// does consume more memory, however.  The updateLock ensures that only
+// (1) goroutine will ever be updating the map at anytime.
+type multiCache struct {
+	basicCache
+}
+
+// fetchPair uses the atomic reference to the keys map and attempts
+// to fetch the key from the cache.
+func (cache *multiCache) fetchPair(keyID string) (pair Pair, ok bool) {
+	pairs, ok := cache.load().(map[string]Pair)
+	if ok {
+		pair, ok = pairs[keyID]
+	}
+
+	return
+}
+
+// copyPairs creates a copy of the current key cache.  If no keys are present
+// yet, this method returns a non-nil empty map.
+func (cache *multiCache) copyPairs() map[string]Pair {
+	pairs, _ := cache.load().(map[string]Pair)
+
+	// make the capacity 1 larger, since this method is almost always
+	// going to be invoked prior to doing a copy-on-write update.
+	newPairs := make(map[string]Pair, len(pairs)+1)
+
+	for keyID, pair := range pairs {
+		newPairs[keyID] = pair
+	}
+
+	return newPairs
+}
+
+func (cache *multiCache) ResolveKey(keyID string) (pair Pair, err error) {
+	var ok bool
+	pair, ok = cache.fetchPair(keyID)
+	if !ok {
+		cache.update(func() {
+			pair, ok = cache.fetchPair(keyID)
+			if !ok {
+				pair, err = cache.delegate.ResolveKey(keyID)
+				if err == nil {
+					newPairs := cache.copyPairs()
+					newPairs[keyID] = pair
+					cache.store(newPairs)
+				}
+			}
+		})
+	}
+
+	return
+}
+
+func (cache *multiCache) UpdateKeys() (count int, errors []error) {
+	if existingPairs, ok := cache.load().(map[string]Pair); ok {
+		count = len(existingPairs)
+		cache.update(func() {
+			newCount := 0
+			newPairs := make(map[string]Pair, len(existingPairs))
+			for keyID, oldPair := range existingPairs {
+				if newPair, err := cache.delegate.ResolveKey(keyID); err == nil {
+					newCount++
+					newPairs[keyID] = newPair
+				} else {
+					// keep the old key in the event of an error
+					newPairs[keyID] = oldPair
+					errors = append(errors, err)
+				}
+			}
+
+			// small optimization: don't bother doing the atomic swap
+			// if every key operation failed
+			if newCount > 0 {
+				cache.store(newPairs)
+			}
+		})
+	}
+
+	return
+}
+
+// NewUpdater conditionally creates a Runnable which will update the keys in
+// the given resolver on the configured updateInterval.  If both (1) the
+// updateInterval is positive, and (2) resolver implements Cache, then this
+// method returns a non-nil function that will spawn a goroutine to update
+// the cache in the background.  Otherwise, this method returns nil.
+func UpdateAtInterval(updateInterval time.Duration, resolver Resolver, waitgroup *sync.WaitGroup, shutdown <-chan struct{}) {
```

Comment:
This is almost what we want.  Some considerations:

- We want to use a `context` together with a cancelation function, which is a common pattern.  The reason is because the act of updating keys may itself be an expensive network operation which we'll want to be able to cancel or time out.
- We don't need a wait group.  We're not allowing a client to wait at a barrier.  Rather, we just want a glorified ticker.